File: cookbooks/temporal-counterfactual-replay/01-fraud-decision-audit.py
"""
Fraud decision audit — what did the world model show when the transfer was approved?

Pattern: a fraud-risk system approves a transfer based on the customer's
state at that moment. Days later, the customer disputes the transfer,
or new evidence emerges that the customer's account had been compromised.
The auditor needs to reconstruct EXACTLY what data the approval saw.

`AS OF seq <param>` returns the world model at any past WAL sequence —
deterministic, bit-for-bit replayable, forever.

Setup: an Account starts clean, then is updated as new fraud signals
arrive. The approval was recorded at seq N. Replay replays.
"""

import shutil
import tempfile

from arcflow import ArcFlow

def main():
    data_dir = tempfile.mkdtemp(prefix="arcflow-fraud-audit-")
    db = ArcFlow(data_dir)

    # Seq 1: Account opens.
    db.execute("""
        CREATE (a:Account {
            id: 'ACC-789',
            holder: 'Alice',
            risk_score: 0.10,
            verified: true
        })
    """)
    print("seq 1: account opened (risk_score=0.10, verified=true)")

    # Seq 2: Approval recorded — at this point in time, risk was 0.10.
    db.execute("""
        CREATE (t:Transfer {id: 'TX-001', amount: 50000, target: 'EXT-9942'})
        WITH t
        MATCH (a:Account {id: 'ACC-789'})
        CREATE (a)-[:AUTHORIZED]->(t)
    """)
    approval_seq = 2  # the seq the transfer was approved at
    print(f"seq {approval_seq}: transfer TX-001 approved against this state")

    # Seq 3: An anti-fraud signal arrives 4 hours later.
    db.execute("""
        MATCH (a:Account {id: 'ACC-789'})
        SET a.risk_score = 0.85
    """)
    print("seq 3: risk score elevated to 0.85 (post-approval signal)")

    # Seq 4: Account flagged for review.
    db.execute("""
        MATCH (a:Account {id: 'ACC-789'})
        SET a.verified = false
    """)
    print("seq 4: verification revoked")

    # The auditor's question: "What did the approval see?"
    print(f"\nAS OF seq {approval_seq} — state visible to the approval system:")
    for r in db.execute(
        "MATCH (a:Account {id: 'ACC-789'}) AS OF seq $s "
        "RETURN a.id AS id, a.risk_score AS risk, a.verified AS verified",
        params={"s": approval_seq},
    ):
        print(f"  {dict(r)}")

    print("\nCurrent state — what the system shows now:")
    for r in db.execute(
        "MATCH (a:Account {id: 'ACC-789'}) "
        "RETURN a.id AS id, a.risk_score AS risk, a.verified AS verified"
    ):
        print(f"  {dict(r)}")

    print("\nObservation: the approval saw risk=0.10, verified=true. The current")
    print("post-fact state shows risk=0.85, verified=false. The approval was made")
    print("on the data available at seq 2 — that's the auditable answer, forever.")

    db.close()
    shutil.rmtree(data_dir, ignore_errors=True)


if __name__ == "__main__":
    main()

File: cookbooks/temporal-counterfactual-replay/rust/src/bin/01-fraud-decision-audit.rs
//! Step 01 (Rust) — fraud decision audit via AS OF seq.
//!
//! A customer's risk score evolves over a short timeline. A transfer
//! is approved at seq N reading the risk-score visible at that
//! moment. Later the risk-score is updated. The auditor asks: "what
//! score was visible at decision time?"

use anyhow::Result;

fn main() -> Result<()> {
    let db = arcflow_sdk::open_concurrent();

    // seq 1: create the customer with initial risk_score
    db.execute(
        "CREATE (:Customer {customer_id: 'C-001', risk_score: 0.21, \
            _observation_class: 'observed', _confidence: 1.0, \
            _source: 'underwriting_v1'})",
    )?;
    println!("seq 1: customer C-001 created with risk_score=0.21");

    // seq 2: record the transfer decision (uses risk_score visible at seq 1)
    db.execute(
        "CREATE (:Transfer {transfer_id: 'T-9381', customer_id: 'C-001', \
            amount: 12500, decision: 'APPROVED', \
            _observation_class: 'observed', _confidence: 1.0, \
            _source: 'fraud_engine_v1'})",
    )?;
    println!("seq 2: transfer T-9381 APPROVED");

    // seq 3: risk score updated (e.g. new pattern recognition fires)
    db.execute(
        "MATCH (c:Customer {customer_id: 'C-001'}) SET c.risk_score = 0.82",
    )?;
    println!("seq 3: risk_score updated to 0.82 (post-decision)");

    // The audit query: same MATCH, AS OF different seqs
    println!("\n--- Audit: what was risk_score at decision time? ---");
    for seq in [1u64, 2, 3] {
        let q = format!(
            "MATCH (c:Customer {{customer_id: 'C-001'}}) \
             AS OF seq {seq} \
             RETURN c.risk_score AS risk"
        );
        for row in db.execute(&q)?.rows {
            println!("  AS OF seq {seq}: risk_score = {}",
                row.get("risk").cloned().unwrap_or_default());
        }
    }

    println!("\n--- Verdict ---");
    println!("  At seq 2 (decision time): risk_score = 0.21 — APPROVAL justified.");
    println!("  The update at seq 3 is post-decision; using it would be look-ahead.");

    Ok(())
}
